contracts/BoostedMasterChefJoe.sol

```diff
@@ -0,0 +1,480 @@
+    struct PoolInfo {
+        IERC20 lpToken;
+        uint256 accJoePerShare;
+        uint256 lastRewardTimestamp;
+        uint256 allocPoint;
+        IRewarder rewarder;
+        // The sum of all veJoe held by users participating in this farm
+        // This value is updated when
+        // - A user enter/leaves a farm
+        // - A user claims veJOE
+        // - A user unstakes JOE
+        uint256 totalVeJoe;
+        // The total boosted `amount` on the farm
+        // This is the sum of all users boosted amounts in the farm. Updated when
+        // totalVeJoe is updated
+        // This is used instead of the usual `lpToken.balanceOf(address(this))`
+        uint256 totalBoostedAmount;
+        // Even though it seems redundant to keep track of two variables here
+        // the `totalBoostedAmount` means we don't need to iterate over all
+        // depositors to calculate the pending reward.
+    }
+
+    /// @notice Address of MCJV2 contract
+    IMasterChefJoe public immutable MASTER_CHEF_V2;
+    /// @notice Address of JOE contract
+    IERC20 public immutable JOE;
+    /// @notice Address of veJOE contract
+    IERC20 public immutable VEJOE;
+    /// @notice The index of BMCJ master pool in MCJV2
+    uint256 public immutable MASTER_PID;
+    /// @notice Info of each BMCJ pool
+    PoolInfo[] public poolInfo;
+    // Set of all LP tokens that have been added as pools
+    EnumerableSet.AddressSet private lpTokens;
+    /// @notice Info of each user that stakes LP tokens
+    mapping(uint256 => mapping(address => UserInfo)) public userInfo;
+    /// @dev Total allocation points. Must be the sum of all allocation points in all pools
+    uint256 public totalAllocPoint;
+    uint256 private constant ACC_TOKEN_PRECISION = 1e18;
+    /// @dev A maximum scaling factor applied to boosted amounts
+    uint256 public maxBoostFactor = 1500;
+    /// @dev Amount of claimable Joe the user has, this is required as we
+    /// need to update rewardDebt after a token operation but we don't
+    /// want to send a reward at this point. This amount gets added onto
+    /// the pending amount when a user claims
+    mapping(uint256 => mapping(address => uint256)) public claimableJoe;
+
+    event Add(uint256 indexed pid, uint256 allocPoint, IERC20 indexed lpToken, IRewarder indexed rewarder);
+    event Set(uint256 indexed pid, uint256 allocPoint, IRewarder indexed rewarder, bool overwrite);
+    event Deposit(address indexed user, uint256 indexed pid, uint256 amount);
+    event Withdraw(address indexed user, uint256 indexed pid, uint256 amount);
+    event UpdatePool(uint256 indexed pid, uint256 lastRewardTimestamp, uint256 lpSupply, uint256 accJoePerShare);
+    event Harvest(address indexed user, uint256 indexed pid, uint256 amount);
+    event EmergencyWithdraw(address indexed user, uint256 indexed pid, uint256 amount);
+    event Init();
+    event MaxBoostFactorChanged(uint256 maxBoostFactor);
```

Comment:
Yeah I don't think it is possible to change maxBoostFactor as it would require iterating over every user in the pool to update their boostedLiquidity, will remove this event.

Reply:
I don't think we can use immutable/constant now as I made the contracts upgrade able.